File: fushi/test/reader/restore_reanchor_transient_viewport_behavior_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:fushi/src/reader/reader_pagination_scripts.dart';
import 'package:fushi/src/reader/reader_study_unit_script.dart';

/// BUG-2652：iOS 上先以分页开书、再在书内切到滚动模式，落点每次都被钉回章首。
///
/// 书内切模式是在**同一个 WKWebView** 里原地重载本章。连续 shell 的恢复把滚动写对了
/// （charOffset 663 → scrollX -1047），但随后几帧原生侧会让 scrollX / scrollY 瞬时读成
/// 0（iOS 模拟器探针：t=3ms 写入 -1047，+16ms 读到 0，期间没有任何 JS 写入，约 40ms 时
/// 滚动树里仍是 -1047）。恰在这个窗口里有两次重锚采样首字锚 = 章首，随后
/// `scrollToChapterStart()`：
///  1. `setChromeInsets()` 重发了已经烘焙进引擎配置的同一组 inset——没有重排要补偿，
///     却照样采样重锚；
///  2. TODO-718 的恢复完成重锚（`beginUiScaleReanchor`）现场采样视口，而不是用恢复
///     自己的锚。
/// 新开书是全新 WebView，没有这个瞬时态，所以用户「退出重进就好」。
///
/// 这里在 `flutter test` 内用 Node 真执行分页 / 连续两个 shell 对象，采样器恒答
/// 「章首」模拟未落定的视口，断言：inset 无变化时不采样、不重锚；有变化时照旧；
/// `beginRestoreReanchor` 取恢复锚（含句尾锚）不采样，无精确锚时退回采样。撤掉修复，
/// Node 断言失败、本 Dart 守卫转红。没有 node 的环境自动 skip。
void main() {
  test(
    'a re-anchor right after a restore never samples the unsettled viewport',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
          'node not found on PATH; skipping JS behavior execution',
        );
        return;
      }

      final File jsTest = File(
        'test/reader/restore_reanchor_transient_viewport_behavior_test.js',
      );
      expect(
        jsTest.existsSync(),
        isTrue,
        reason: 'behavior harness ${jsTest.path} must exist',
      );

      final String payload = jsonEncode(<String, String>{
        'paged': ReaderPaginationScripts.paginatedShellSource(),
        'continuous': ReaderPaginationScripts.continuousShellSource(),
        'studyUnits': kStudyUnitJs,
      });
      final Directory temp = Directory.systemTemp.createTempSync(
        'fushi-restore-reanchor-transient-',
      );
      final File payloadFile = File('${temp.path}/payload.json')
        ..writeAsStringSync(payload);
      late final ProcessResult result;
      try {
        result = await Process.run(
          nodeExe,
          <String>[jsTest.path, payloadFile.path],
          workingDirectory: Directory.current.path,
          stdoutEncoding: utf8,
          stderrEncoding: utf8,
        );
      } finally {
        temp.deleteSync(recursive: true);
      }

      expect(
        result.exitCode,
        0,
        reason:
            'restore re-anchor JS behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(
        result.stdout.toString(),
        contains('all assertions passed'),
        reason: 'behavior harness must reach its success marker',
      );
    },
  );
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates = Platform.isWindows
      ? <String>['node.exe', 'node']
      : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}

File: fushi/test/reader/restore_reanchor_transient_viewport_behavior_test.js
// BUG-2652 behavior test: a re-anchor that runs right after a restore must not
// sample the viewport.
//
// Regression context (iOS, 2026-09-25): opening a book in paged mode and
// switching to scroll mode from the in-book settings reloads the chapter in the
// SAME WKWebView. The continuous restore writes the right scroll (e.g. -1047 for
// charOffset 663), but for the next few frames WebKit's native side makes
// scrollX / scrollY read back as 0 (probe on the iOS simulator: -1047 written at
// t=3ms, 0 read at +16ms with no JS write in between, -1047 again in the
// scrolling tree at ~40ms). Two re-anchors ran inside that window and sampled
// getFirstVisibleCharOffset() == chapter start, then scrollToChapterStart() —
// the reader landed on the chapter start every time (8/8 switches), while a
// fresh open (new WebView) never shows the transient:
//   1. setChromeInsets() re-sent the insets that were already baked into the
//      engine config — no reflow to compensate, yet it sampled and re-anchored;
//   2. the TODO-718 restore re-anchor (beginUiScaleReanchor) sampled the
//      viewport instead of using the restore's own anchor.
//
// This harness instantiates the real paged and continuous shell objects
// (extracted from reader_pagination_scripts.dart via the Dart driver) with a
// viewport sampler that always answers "chapter start", and asserts:
//   A. setChromeInsets with unchanged insets + image box neither samples nor
//      re-anchors (both shells), and leaves the paged metrics cache alone;
//   B. setChromeInsets with a real inset change still samples and re-anchors;
//   C. beginRestoreReanchor takes the registered char anchor (and the sentence
//      end) without sampling, and commitUiScaleReanchor scrolls to it;
//   D. without a precise char anchor (progress restore) it falls back to the
//      sampling begin, and the UI-scale path still passes no sentence end.
//
// Run: node fushi/test/reader/restore_reanchor_transient_viewport_behavior_test.js <payload.json>
// (driven from restore_reanchor_transient_viewport_behavior_test.dart inside `flutter test`).

const assert = require('assert');
const fs = require('fs');

const data = JSON.parse(fs.readFileSync(process.argv[2], 'utf8'));

function objectLiteral(source) {
  const marker = 'window.fushiReader = {';
  const start = source.indexOf(marker);
  assert.ok(start >= 0, 'fushiReader object missing');
  const brace = source.indexOf('{', start);
  const end = source.indexOf('\n};', brace);
  assert.ok(end >= 0, 'fushiReader object terminator missing');
  return source.slice(brace, end + 2);
}

const IMAGE_BOX = { w: 300, h: 700 };

function instantiate(shellSource) {
  const vars = new Map();
  const rootStyle = {
    setProperty(k, v) { vars.set(k, String(v)); },
    getPropertyValue(k) { return vars.has(k) ? vars.get(k) : ''; },
  };
  const window = {
    scrollX: -1047, scrollY: 0, innerWidth: 402, innerHeight: 874,
    CSS: {}, Highlight: function() {},
  };
  const cs = {
    writingMode: 'vertical-rl', columnWidth: 'auto',
    paddingTop: '0px', paddingBottom: '0px', paddingLeft: '0px', paddingRight: '0px',
  };
  const getComputedStyle = () => cs;
  window.getComputedStyle = getComputedStyle;
  const document = {
    documentElement: { style: rootStyle, clientHeight: 874, clientWidth: 402 },
    scrollingElement: null,
    body: { clientWidth: 402, clientHeight: 874 },
    caretRangeFromPoint: () => null,
  };
  const Node = { TEXT_NODE: 3 };
  new Function('window', data.studyUnits)(window);
  const C = { perfTraceEnabled: false };
  const factory = new Function(
    'window', 'document', 'C', 'global', 'CSS', 'Highlight', 'getComputedStyle', 'Node',
    'window.fushiReader = ' + objectLiteral(shellSource) + '; return window.fushiReader;'
  );
  const reader = factory(window, document, C, {}, window.CSS, window.Highlight, getComputedStyle, Node);

  const calls = { sample: 0, scrollTo: [] };
  // The unsettled viewport: every sample answers "chapter start".
  reader.getFirstVisibleCharOffset = () => { calls.sample++; return 0; };
  reader.scrollToCharOffset = function() { calls.scrollTo.push(Array.from(arguments)); };
  reader._reanchorFrame = (fn) => fn();
  reader._imageMaxBox = () => IMAGE_BOX;
  // Paged-only geometry used for the hint; harmless on the continuous shell.
  reader.getScrollContext = () => ({ vertical: true, pageSize: 874 });
  reader.getPagePosition = () => 874;

  // What initialize() leaves behind: the insets and image box of this layout.
  rootStyle.setProperty('--chrome-top-inset', '62px');
  rootStyle.setProperty('--chrome-bottom-inset', '34px');
  rootStyle.setProperty('--fushi-image-max-width', IMAGE_BOX.w + 'px');
  rootStyle.setProperty('--fushi-image-max-height', IMAGE_BOX.h + 'px');
  return { reader, calls, rootStyle };
}

for (const [label, source] of [['paged', data.paged], ['continuous', data.continuous]]) {
  // A. unchanged insets → no sample, no re-anchor, metrics cache untouched.
  {
    const { reader, calls } = instantiate(source);
    const metrics = { sentinel: true };
    reader.paginationMetrics = metrics;
    reader.setChromeInsets(62, 34);
    assert.strictEqual(calls.sample, 0,
      label + ': unchanged insets must not sample the (possibly unsettled) viewport');
    assert.strictEqual(calls.scrollTo.length, 0,
      label + ': unchanged insets must not re-anchor (scrollToCharOffset(0) = chapter start)');
    assert.notStrictEqual(reader._reanchorPending, true,
      label + ': unchanged insets must not raise the re-anchor flag');
    assert.strictEqual(reader.paginationMetrics, metrics,
      label + ': unchanged insets must keep the pagination metrics cache');
  }

  // A'. same insets but a stale image box is still a layout change.
  {
    const { reader, calls, rootStyle } = instantiate(source);
    rootStyle.setProperty('--fushi-image-max-height', '640px');
    reader.setChromeInsets(62, 34);
    assert.strictEqual(calls.sample, 1,
      label + ': a changed image box reflows, so the re-anchor must still sample');
    assert.strictEqual(rootStyle.getPropertyValue('--fushi-image-max-height'), IMAGE_BOX.h + 'px',
      label + ': the image box must be re-derived');
  }

  // B. a real inset change keeps the existing re-anchor behaviour.
  {
    const { reader, calls, rootStyle } = instantiate(source);
    reader.setChromeInsets(62, 90);
    assert.strictEqual(calls.sample, 1, label + ': a changed inset must sample the anchor');
    assert.strictEqual(calls.scrollTo.length, 1, label + ': a changed inset must re-anchor');
    assert.strictEqual(rootStyle.getPropertyValue('--chrome-bottom-inset'), '90px',
      label + ': the new inset must be applied');
    assert.notStrictEqual(reader._reanchorPending, true,
      label + ': the re-anchor frame must clear the flag');
  }
}

// C. restore re-anchor takes the restore's own anchor.
{
  const { reader, calls } = instantiate(data.continuous);
  reader.registerImageLateAnchor({ charOffset: 663, endCharOffset: -1 });
  assert.strictEqual(reader.beginRestoreReanchor(), 663,
    'restore re-anchor must return the restore char anchor');
  assert.strictEqual(calls.sample, 0,
    'restore re-anchor must not sample the unsettled viewport');
  assert.strictEqual(reader._reanchorPending, true,
    'restore re-anchor must hold the flag until commit');
  assert.strictEqual(reader.commitUiScaleReanchor(), true, 'commit must run');
  assert.deepStrictEqual(calls.scrollTo[0].slice(0, 2), [663, -1],
    'commit must scroll back to the restore anchor');
  assert.notStrictEqual(reader._reanchorPending, true, 'commit must clear the flag');

  // Favourite-sentence restore keeps its sentence end (BUG-461 whole-sentence alignment).
  reader.registerImageLateAnchor({ charOffset: 100, endCharOffset: 140 });
  assert.strictEqual(reader.beginRestoreReanchor(), 100);
  reader.commitUiScaleReanchor();
  assert.deepStrictEqual(calls.scrollTo[1].slice(0, 2), [100, 140],
    'commit must carry the restore sentence end');

  // An in-flight re-anchor keeps ownership.
  reader._reanchorPending = true;
  reader.registerImageLateAnchor({ charOffset: 200 });
  assert.strictEqual(reader.beginRestoreReanchor(), -1,
    'an in-flight re-anchor must keep ownership');
  reader._reanchorPending = false;
}

// D. no precise anchor → sampling fallback; UI-scale path passes no sentence end.
{
  const { reader, calls } = instantiate(data.continuous);
  reader.registerImageLateAnchor({ progress: 0.4 });
  assert.strictEqual(reader.beginRestoreReanchor(), 0,
    'a progress restore falls back to the sampling begin');
  assert.strictEqual(calls.sample, 1, 'the fallback samples once');
  reader.commitUiScaleReanchor();

  reader.clearImageLateAnchor();
  reader.beginUiScaleReanchor();
  reader.commitUiScaleReanchor();
  assert.strictEqual(calls.scrollTo[calls.scrollTo.length - 1][1], undefined,
    'the UI-scale re-anchor must not inherit a sentence end');
}

console.log('all assertions passed');
